ext/simplexml/simplexml.c

```diff
@@ -2378,7 +2378,7 @@ PHP_METHOD(SimpleXMLElement, __construct)
 	PHP_LIBXML_RESTORE_GLOBALS(read_file_or_memory);
 
 	if (!docp) {
-		((php_libxml_node_object *)sxe)->document = NULL;
+		php_libxml_decrement_node_ptr((php_libxml_node_object *)sxe);
```

Comment:
I'm not entirely sure that calling __construct on a working SimpleXMLElement (SXE) which then fails should cause the SXE to no longer work. I'd move this decrement code to the happy path below.
Furthermore, there's still leaks here in the success case because the string duplication of `nsprefix` is not freed, and the fact you don't decrement the document reference.
Also I bet that iteration and xpath data etc should be reset as well on the happy path, so I think you should factor out part of the `sxe_object_free_storage` function and call that factored out code on the happy path here.